_source/_posts/2022-02-23-angular-async-config.md

```diff
@@ -0,0 +1,349 @@
+
+You'll add a new file and create an interface that matches the response, as well as a class to hold the config:
+
+```ts
+export interface AuthConfig {
+  issuer: string;
+  clientId: string;
+}
+
+export class OktaAuthConfig {
+  constructor(public config: AuthConfig) { }
+}
+```
+
+Edit the `src/main.ts` to provide the `OktaAuthConfig` class instead
+
+```ts
+import { OktaAuthConfig } from './app/auth-config';
+
+fetch('http://{yourApiUri}/config').then(async res => {
+  const authConfig = new OktaAuthConfig(await res.json());
+
+  platformBrowserDynamic([
+    { provide: OktaAuthConfig, useValue: authConfig }
+  ]).bootstrapModule(AppModule)
+  .catch(err => console.error(err));
+})
+```
+
+In the `AppModule` you can provide the `OKTA_CONFIG` needed to integrate with Okta by accessing `OktaAuthConfig`:
+
+```ts
+@NgModule({
+  declarations: [ AppComponent, ProfileComponent ],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    OktaAuthModule
+  ],
+  providers: [
+    {
+      provide: OKTA_CONFIG,
+      deps: [OktaAuthConfig],
+      useFactory: (oktaConfig: OktaAuthConfig) => ({
+        oktaAuth: new OktaAuth({
+          ...oktaConfig.config,
+          redirectUri: window.location.origin + '/login/callback'
+        })
+      })
+    }
+  ]
+})
+export class AppModule { }
+```
+
+You can now load a config from an API and use the app's location.
+
+You might be asking yourself, "Isn't there an `APP_INITIALIZER` token or something we can use instead"? Well, yes, there is an [`APP_INITIALIZER` token](https://angular.io/api/core/APP_INITIALIZER) for executing initialization functions that complete before application initialization completes. However, in our case, we need the auth configuration _in order_ to initialize. So, we need to finish loading the configuration before initializing the app, which we can do when bootstrapping.
+
+**Summary of _loading configuration from an external API_:**
+
+Load configuration from an API and provide the config to the application. Depending on your needs, the configuration loading may occur during bootstrapping or via `APP_INITIALIZER`.
+
+**When to use:**
+
+* You want the configuration to load at runtime instead of being baked into the code
+* Your configuration properties include private information that you don't want to commit to source control
+
+**Best use cases:**
+
+* You have different configurations for staging and prod and use release-promotion style deployment processes
+* Your configuration changes frequently or often enough where building and deploying the app is not feasible
+
+**Watch out for:**
+
+* Configuration errors or network blips - your app **will not run** since it's dependent on the external API.
+* Anything that can decrease application load speed such as overly large configuration response, calling too many endpoints, or slow server response
+* Potentially harder to verify and test since configuration may change.
```

Comment:
```suggestion
* Potential challenges with verification and testing, since configuration may change.
```
consistency of bullet format. Each bullet should form a complete sentence with "Watch out for:..." when you read it aloud in your head...